presto-manager-common/src/main/java/com/teradata/prestomanager/common/NodeSet.java

```diff
@@ -0,0 +1,113 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.teradata.prestomanager.common;
+
+import java.net.URI;
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+public class NodeSet
+{
+    private static NodeSet nodeSetInstance = new NodeSet();
+
+    public static Set<Agent> nodeSet = ConcurrentHashMap.newKeySet();
+
+    public static NodeSet getInstance()
+    {
+        return nodeSetInstance;
+    }
+
+    public void addAgent(URI uri, NodeRole nodeRole)
+    {
+        nodeSet.add(new Agent(uri, nodeRole));
+    }
+
+    public void removeAgent(Integer id)
+    {
+        nodeSet.removeIf(agent -> agent.getId().equals(id));
+    }
+
+    public URI getUriById(Integer id)
+    {
+        return nodeSet.stream().filter(agent -> agent.getId().equals(id)).map(Agent::getUri).findAny().orElse(null);
+    }
+
+    public Collection<URI> getAllNodes()
+    {
+        return nodeSet.stream().map(Agent::getUri).collect(Collectors.toList());
+    }
+
+    public Collection<URI> getCoordinator()
+    {
+        return nodeSet.stream().filter(agent -> agent.getNodeRole() == NodeRole.COORDINATOR).map(Agent::getUri).collect(Collectors.toList());
+    }
+
+    public Collection<URI> getWorkers()
+    {
+        return nodeSet.stream().filter(agent -> agent.getNodeRole() == NodeRole.WORKER).map(Agent::getUri).collect(Collectors.toList());
+    }
+
+    public enum NodeRole
+    {
+        WORKER, COORDINATOR
+    }
+
+    //TODO: Add method to generate ID
+    public static class Agent
+    {
+        private URI uri;
+        private static final AtomicInteger count = new AtomicInteger(0);
+        private int id;
+        private NodeRole nodeRole;
+
+        private Agent(URI uri, NodeRole nodeRole)
+        {
+            this.uri = requireNonNull(uri, "uri is null");
+            this.nodeRole = requireNonNull(nodeRole, "nodeRole is null");
+
+            this.id = count.incrementAndGet();
+        }
+
+        private URI getUri()
+        {
+            return uri;
+        }
+
+        private Integer getId()
+        {
+            return id;
+        }
+
+        private NodeRole getNodeRole() {return nodeRole;}
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            return (obj instanceof Agent) && (getId().equals(((Agent) obj).getId()));
```

Comment:
Equals is based on the id field alone because that prevents a `Set` from containing the same Agent twice.